editor/shader_create_dialog.cpp

```diff
@@ -0,0 +1,653 @@
+	path_error_label->set_text("- " + p_msg);
+	if (valid) {
+		path_error_label->add_theme_color_override("font_color", gc->get_theme_color("success_color", "Editor"));
+	} else {
+		path_error_label->add_theme_color_override("font_color", gc->get_theme_color("error_color", "Editor"));
+	}
+}
+
+void ShaderCreateDialog::_update_dialog() {
+	bool shader_ok = true;
+
+	if (!is_built_in && !is_path_valid) {
+		_msg_script_valid(false, TTR("Invalid path."));
+		shader_ok = false;
+	}
+	if (shader_ok) {
+		_msg_script_valid(true, TTR("Shader path/name is valid."));
+	}
+	if (!built_in_enabled) {
+		internal->set_pressed(false);
+	}
+
+	if (is_built_in) {
+		file_path->set_editable(false);
+		path_button->set_disabled(true);
+		re_check_path = true;
+	} else {
+		file_path->set_editable(true);
+		path_button->set_disabled(false);
+		if (re_check_path) {
+			re_check_path = false;
+			_path_changed(file_path->get_text());
+		}
+	}
+
+	internal->set_disabled(!built_in_enabled);
+
+	builtin_warning_label->set_visible(is_built_in);
+
+	if (is_built_in) {
+		get_ok_button()->set_text(TTR("Create"));
+		_msg_path_valid(true, TTR("Built-in script (into scene file)."));
+	} else if (is_new_shader_created) {
+		get_ok_button()->set_text(TTR("Create"));
+		if (is_path_valid) {
+			_msg_path_valid(true, TTR("Will create a new shader file."));
+		}
+	} else if (load_enabled) {
+		get_ok_button()->set_text(TTR("Load"));
+		if (is_path_valid) {
+			_msg_path_valid(true, TTR("Will load an existing shader file."));
+		}
+	} else {
+		get_ok_button()->set_text(TTR("Create"));
+		_msg_path_valid(false, TTR("Shader file already exists."));
+
+		shader_ok = false;
+	}
+
+	get_ok_button()->set_disabled(!shader_ok);
+
+	Callable entered_call = callable_mp(this, &ShaderCreateDialog::_path_submitted);
+	if (shader_ok) {
+		if (!file_path->is_connected("text_submitted", entered_call)) {
+			file_path->connect("text_submitted", entered_call);
+		}
+	} else if (file_path->is_connected("text_submitted", entered_call)) {
+		file_path->disconnect("text_submitted", entered_call);
+	}
+}
+
+void ShaderCreateDialog::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("config", "path", "built_in_enabled", "load_enabled"), &ShaderCreateDialog::config, DEFVAL(true), DEFVAL(true));
+
+	ADD_SIGNAL(MethodInfo("shader_created", PropertyInfo(Variant::OBJECT, "shader", PROPERTY_HINT_RESOURCE_TYPE, "Shader")));
+}
+
+ShaderCreateDialog::ShaderCreateDialog() {
+	_update_language_info();
+
+	// Main Controls.
+
+	gc = memnew(GridContainer);
+	gc->set_columns(2);
+
+	// Error Fields.
+
+	VBoxContainer *vb = memnew(VBoxContainer);
+
+	error_label = memnew(Label);
+	vb->add_child(error_label);
+
+	path_error_label = memnew(Label);
+	vb->add_child(path_error_label);
+
+	builtin_warning_label = memnew(Label);
+	builtin_warning_label->set_text(
+			TTR("Note: Built-in shaders have some limitations and can't be edited using an external editor."));
```

Comment:
The part about limitations isn't really true, built-in shaders have exactly the same functionality as file shaders (unlike scripts). The message should be reworded to only mention external editors.

Reply:
Fixed